File: algoexpert/easy/validate_subsequence/python/validate-subsequence.py
# solution1
# Time: O(n)
# Space: O(1)
def isValidSubsequence1(array, sequence):
    arrayIdx = 0
    seqIdx = 0
    while arrayIdx < len(array) and seqIdx < len(sequence):
        if array[arrayIdx] == sequence[seqIdx]:
            seqIdx += 1
        arrayIdx += 1
    return seqIdx == len(sequence)


# solution2
# Time: O(n)
# Space: O(1)
def isValidSubsequence2(array, sequence):
    seqIdx = 0
    for value in array:
        if seqIdx == len(sequence):
            break
        if sequence[seqIdx] == value:
            seqIdx += 1
    return seqIdx == len(sequence)


result1 = isValidSubsequence1([5, 6, 2, 3, 4], [6, 4])
result2 = isValidSubsequence2([5, 6, 2, 3, 4], [6, 4])
print(result1, result2)

File: algoexpert/easy/validate_subsequence/typescript/validate-subsequence.ts
// solution1
function isValidSubsequence1(array: number[], sequence: number[]): boolean {
  let arrayIdx = 0;
  let seqIdx = 0;
  while (arrayIdx < array.length && seqIdx < sequence.length) {
    if (array[arrayIdx] === sequence[seqIdx]) {
      seqIdx++;
    }
    arrayIdx++;
  }
  return seqIdx === sequence.length;
}

// solution2
function isValidSubsequence2(array: number[], sequence: number[]): boolean {
  let seqIdx = 0;
  for (let value of array) {
    if (seqIdx === sequence.length) {
      break;
    }
    if (sequence[seqIdx] === value) {
      seqIdx++;
    }
  }
  return seqIdx === sequence.length;
}

const result1 = isValidSubsequence1([4, 5, 2, 3, 4, 6], [5, 4, 6]);
const result2 = isValidSubsequence2([4, 5, 2, 3, 4, 6], [5, 4, 6]);
console.table({ result1, result2 });
